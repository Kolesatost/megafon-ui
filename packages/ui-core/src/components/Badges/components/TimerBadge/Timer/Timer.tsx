import * as React from 'react';
import { cnCreate, filterDataAttrs } from '@megafon/ui-helpers';
import format from 'date-fns/format';
import ruLocale from 'date-fns/locale/ru';
import * as PropTypes from 'prop-types';

const SECONDS_IN_DAY = 86400;
const SECONDS_IN_HOUR = 3600;
const SECONDS_IN_MINUTE = 60;

const formatDate = (date: Date, pattern: string) => format(date, pattern, { locale: ruLocale });

export const getCountdownText = (remainingTime = 0): string => {
    const isMoreHourAndLessDay = remainingTime >= SECONDS_IN_HOUR && remainingTime < SECONDS_IN_DAY;

    const truncTime = (divider = 1) => Math.trunc(remainingTime / divider);

    switch (true) {
        case remainingTime <= 0: {
            return '';
        }
        case remainingTime < SECONDS_IN_MINUTE: {
            const seconds = truncTime();

            return `${seconds} сек`;
        }
        case remainingTime < SECONDS_IN_HOUR: {
            const minutes = truncTime(SECONDS_IN_MINUTE);

            return `${minutes} мин`;
        }
        case isMoreHourAndLessDay: {
            const hours = truncTime(SECONDS_IN_HOUR);
            const secondsLeft = remainingTime - hours * SECONDS_IN_HOUR;
            const minutes = Math.trunc(secondsLeft / SECONDS_IN_MINUTE);

            return minutes ? `${hours} ч ${minutes} мин` : `${hours} ч`;
        }
        case remainingTime >= SECONDS_IN_DAY: {
            const days = truncTime(SECONDS_IN_DAY);
            const secondsLeft = remainingTime - days * SECONDS_IN_DAY;
            const hours = Math.trunc(secondsLeft / SECONDS_IN_HOUR);

            return hours ? `${days} дн ${hours} ч` : `${days} дн`;
        }
        default: {
            return '';
        }
    }
};

interface ITimerProps {
    actualRemainingTime: number;
    expirationDate: Date;
    showCountdown: boolean;
    additionalText?: string;
    dataAttrs?: {
        root?: Record<string, string>;
    };
}

const cn = cnCreate('mfui-timer');
const Timer: React.FC<ITimerProps> = ({
    actualRemainingTime,
    expirationDate,
    showCountdown,
    additionalText,
    dataAttrs,
}) => (
    <div className={cn()} {...filterDataAttrs(dataAttrs?.root)}>
        {additionalText && <span>{additionalText} </span>}
        <span>
            {showCountdown ? getCountdownText(actualRemainingTime) : formatDate(expirationDate, 'dd MMMM yyyy')}
        </span>
    </div>
);

Timer.propTypes = {
    expirationDate: PropTypes.instanceOf(Date).isRequired,
    actualRemainingTime: PropTypes.number.isRequired,
    showCountdown: PropTypes.bool.isRequired,
    additionalText: PropTypes.string,
    dataAttrs: PropTypes.shape({
        root: PropTypes.objectOf(PropTypes.string.isRequired),
    }),
};

export default Timer;
