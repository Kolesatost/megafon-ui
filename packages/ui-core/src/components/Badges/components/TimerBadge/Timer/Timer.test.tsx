import * as React from 'react';
import { shallow } from 'enzyme';
import Timer, { getCoundownText } from './Timer';

const defaultDate = new Date('2022-04-25T14:50:15z');

const defaultProps = {
    expirationDate: defaultDate,
    actualRemainingTime: 1000,
    showCountdown: false,
};

const defaultAllProps = {
    dataAttrs: {
        root: { 'data-testid': 'root' },
    },
    additionalText: 'Действительно до',
    expirationDate: defaultDate,
    actualRemainingTime: 2000,
    showCountdown: true,
};

describe('<Timer />', () => {
    describe('snapshots', () => {
        it('renders with require props', () => {
            const wrapper = shallow(<Timer {...defaultProps} />);

            expect(wrapper).toMatchSnapshot();
        });

        it('renders with all props', () => {
            const wrapper = shallow(<Timer {...defaultAllProps} />);

            expect(wrapper).toMatchSnapshot();
        });
    });
});

describe('getCoundownText tests', () => {
    it('should return correct string, if remaining time less than minute', () => {
        const result = getCoundownText(55);

        expect(result).toBe('55 сек');
    });

    it('should return correct string, if remaining time less than hour', () => {
        const result = getCoundownText(3500);

        expect(result).toBe('58 мин');
    });

    it('should return correct string, if remaining time more than hour and less than one day', () => {
        const result = getCoundownText(4000);

        expect(result).toBe('1 ч 6 мин');
    });

    it('should return correct string, if remaining time includes days and hours', () => {
        const result = getCoundownText(94000);

        expect(result).toBe('1 дн 2 ч');
    });

    it('should return correct string, if remaining time includes full days', () => {
        const result = getCoundownText(172800);

        expect(result).toBe('2 дн');
    });

    it('should return empty string without remaining time', () => {
        const result = getCoundownText();

        expect(result).toBe('');
    });
});
