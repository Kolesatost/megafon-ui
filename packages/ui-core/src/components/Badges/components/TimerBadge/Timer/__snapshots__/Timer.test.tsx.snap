// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timer /> snapshots renders with all props 1`] = `
<div
  className="mfui-timer"
  data-testid="root"
>
  <span>
    Действительно до
     
  </span>
  <span>
    33 мин
  </span>
</div>
`;

exports[`<Timer /> snapshots renders with require props 1`] = `
<div
  className="mfui-timer"
>
  <span>
    25 апреля 2022
  </span>
</div>
`;
